apps/src/lab2/views/Lab2Wrapper.tsx

```diff
@@ -62,6 +63,12 @@ const Lab2Wrapper: React.FunctionComponent<Lab2WrapperProps> = ({children}) => {
     }
   }, [isShareView, dispatch]);
 
+  useEffect(() => {
+    window.speechSynthesis.onvoiceschanged = () => {
+      dispatch(setTextToSpeechHasVoices(true));
+    };
+  }, [dispatch]);
```

Comment:
Should this happen in the `TextToSpeech` component so it's only enabled for labs/components that explicitly include it?

Reply:
I was wondering the same, and intended for this first pass to be a bit of a conversation-starter!

I'd been thinking there could be some advantage to having some kind of singleton handling these updates, and then sharing the results amongst active `TextToSpeech` UI components (since there could be none, one, or several on screen at any given time).  In that case, we could maybe wrap the lab with a `TtsManager` the same way we already wrap it with a `DialogManager`.  

But maybe it's fine for each `TextToSpeech` UI component to check the current voice list, hide the UI if there are none yet available, do its own `addEventListener`, and respond to updates.  The sample code [here](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesis#examples) and [here](https://github.com/mdn/dom-examples/blob/a6ee8145d8014e4155348aa19822ea942597b935/web-speech-api/speak-easy-synthesis/script.js) might be helpful.

Reply:
I think having a centralized singleton is a good idea. Had a couple thoughts:

- We could have some non-React, centralized `BrowserTextToSpeech.ts` module, that exports functions like `onTextToSpeechAvailable(callback: () => void)`, `isTextToSpeechAvailable(): boolean`, and `speak(text: string)`. This module can internally handle all the specifics like attaching the DOM listener, managing and picking the appropriate voice, stripping markdown, etc.
- Alternatively or additionally, we could have a wrapper component like you mentioned that uses React Context. This wrapper can track the state of if TTS is available and provide the state to child components. That way, each button doesn't need to attach its own listener, but just get the availability status from the context.

Happy to discuss further and/or prototype these ideas on this branch if it's helpful

Reply:
Great! Just a pushed a new commit that adds the above -
- `BrowserTextToSpeech.ts` is a module that tracks the state and includes the functions as mentioned above
- `BrowserTextToSpeechWrapper.ts` is a React wrapper that components can use to provide TTS functionality via context. Child components (like the Lab2 TextToSpeech button) can also use the exported `useBrowserTextToSpeech()` hook for easy access to TTS availability state and the `speak` function.

Let me know what you think!

Reply:
Looks great.  Also tried it out, and it continues to work well.